### PR TITLE
Added T3/T5 branches for GNN T5 classification

### PR DESCRIPTION
**Pull request by jkguiang**
Added GNN NTuple branches for T5 classification plus some minor fixes to Event.cu

### CODE REVIEW

**Comment by jkguiang** [comment]
Let me know if I should address #249 in this PR as well. I think the fix is easy.

File: code/core/write_sdl_ntuple.cc [hunk 1]
```diff
@@ -782,6 +843,195 @@ void setGnnNtupleMiniDoublet(SDL::Event* event, unsigned int MD)
...
+        short module_side = modulesInGPU.sides[module];
+        float module_drdz = modulesInGPU.drdzs[module];
+        float module_slope = modulesInGPU.slopes[module];
+        // Get deltas for sigma calculation
+        float delta1, delta2;
+        bool is_flat;
+        // Category 1: barrel PS flat
+        if (module_subdet == SDL::Barrel and module_type == SDL::PS and module_side == SDL::Center)
+        {
+            delta1 = inv1;//1.1111f;//0.01;
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:916]: Commented out numbers should be deleted? The first one seems to be the result of `inv1` (so why have it re-written), while the second one is random - at least I can't understand where it comes from...

This comment applies to similar instances below.

File: code/core/write_sdl_ntuple.cc [hunk 2]
```diff
@@ -782,6 +843,195 @@ void setGnnNtupleMiniDoublet(SDL::Event* event, unsigned int MD)
...
+        }
+        // Category 4: endcap PS
+        else if (module_subdet == SDL::Endcap and module_type == SDL::PS)
+        {
+            delta1 = inv1;//1.1111f;//0.01;
+            is_flat = false;
+
+            /* Despite the type of the module layer of the lower module index,
+             * all anchor hits are on the pixel side and all non-anchor hits are
+             * on the strip side! */
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:946]: Is this comment relevant here? If it is, probably its connections should be made clearer?

File: code/core/write_sdl_ntuple.cc [hunk 3]
```diff
@@ -782,6 +843,195 @@ void setGnnNtupleMiniDoublet(SDL::Event* event, unsigned int MD)
...
+        sigmas.push_back(2 * sqrtf((xPrime * delta1) * (xPrime * delta1) + (yPrime * delta2) * (yPrime * delta2)));
+    }
+
+    ana.tx->pushbackToBranch<float>("t3_0_dr", sigmas[0]);
+    ana.tx->pushbackToBranch<float>("t3_2_dr", sigmas[1]);
+    ana.tx->pushbackToBranch<float>("t3_4_dr", sigmas[2]);
+
+    // Constants
+    const float kRinv1GeVf = (2.99792458e-3 * 3.8);
+    const float k2Rinv1GeVf = kRinv1GeVf / 2.;
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:1005]: Should we just include the constants from [here](https://github.com/SegmentLinking/TrackLooper/blob/579b9d08bfb5d774c6ce3727e7625a4620faf1ac/SDL/Constants.h#L78)?

File: code/core/write_sdl_ntuple.cc [hunk 4]
```diff
@@ -782,6 +843,195 @@ void setGnnNtupleMiniDoublet(SDL::Event* event, unsigned int MD)
...
+
+    // Radial distance
+    const float dr = sqrt(pow(hitsInGPU.xs[hit4] - hitsInGPU.xs[hit0], 2) + pow(hitsInGPU.ys[hit4] - hitsInGPU.ys[hit0], 2));
+
+    // Beta angles
+    float betaIn  = __H2F(tripletsInGPU.betaIn[T3]);
+    float betaOut = __H2F(tripletsInGPU.betaOut[T3]);
+
+    // Legacy T4 pt estimate
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:1014]: T4? I guess this is confusing, because we don't have T4s now?

File: code/core/write_sdl_ntuple.cc [hunk 5]
```diff
@@ -1042,7 +1292,7 @@ std::tuple<float, float, float, vector<unsigned int>, vector<unsigned int>> pars
     const float ptAv_out = abs(dr_out * k2Rinv1GeVf / sin((betaIn_out + betaOut_out) / 2.));
 
     // T5 pt is average of the two pt estimates
-    const float pt = (ptAv_in + ptAv_out) / 2.;
+    const float pt = (ptAv_in + ptAv_out) / 2.;     // this is deprecated, c.f. setQuintupletOutputBranches
```
Comment (VourMa) [inline code/core/write_sdl_ntuple.cc:1295]: Not sure I understand the comment. Is this used? If not, maybe we should delete it and use the up-to-date version?

**Review by VourMa — COMMENTED** [review]
Since two new variables are added (are they used anywhere, by the way?), could you please run the checks [here](https://github.com/SegmentLinking/TrackLooper/wiki/Line-Segment-Tracking-Algorithm#debugging) to be sure that everything checks out?

**Comment by sgnoohc** [comment]
general comment.
Is there a reason why not to put idxs to the relevant LS in the t3 branches?
The flattening of the t3 hit positions are OK in principle.
But it puts a danger in creating a lot of branches and lots of duplicate information.
(e.g. if two T3's share a LS, it's going to copy paste same information twice.)

I would rather do something like
T3_LS_idx0
keeping the same branch philosophy as the LS.

File: code/core/write_sdl_ntuple.cc [hunk 6]
```diff
@@ -181,6 +181,30 @@ void createGnnNtupleBranches()
     ana.tx->createBranch<vector<float>>("LS_sim_vz");
     ana.tx->createBranch<vector<int>>("LS_isInTrueTC");
 
+    // T3 branches
+    ana.tx->createBranch<vector<int>>("t5_t3_idx0");
+    ana.tx->createBranch<vector<int>>("t5_t3_idx1");
```
Comment (sgnoohc) [inline code/core/write_sdl_ntuple.cc:186]: parallelism is broken here.
Technically these are t5 branches.
The comment //T3 branches showing right above this is misleading

**Comment by jkguiang** [comment]
> Is there a reason why not to put idxs to the relevant LS in the t3 branches?

We use the three hits stored in the t3 branches for computing Balaji's Chi^2. Storing them directly makes it very easy to use them as inputs to the GNN, rather than requiring nested lookups (first to the LS, then to the MD). I also figure that there is not much "danger" here for now, as this is just for training the GNN.

> I would rather do something like
T3_LS_idx0
keeping the same branch philosophy as the LS.

This is doable if necessary, but would be more of a headache as explained above.

File: code/core/write_sdl_ntuple.cc (hunk 6 above)
Comment (jkguiang) [reply code/core/write_sdl_ntuple.cc:186]: What do you mean by "parallelism is broken"? The comment is a typo and will be fixed with the responses to Manos's comments.

**Pull request closed without merging**